### ISSUE TITLE
[BUG] gpuinflate function occasionally hangs on Maxwell devices and outs of bounds on Pascal or Volta

### ISSUE THREAD

**Issue by isolenta** [rapidsai/cudf#3800]
**Describe the bug**

Attached test uses unmodified gpuinflate kernel sources from cuDF. Test uses gzip-compressed data files as input. Test decompresses input file by standard zlib first and by gpuinflate next. After gpuinflate finished it checks kernel statuses (gpu_inflate_status_s status field) and byte-to-byte compares output result with zlib output. On Maxwell devices gpuinflate kernel occasionally hangs (probably on busy loop inside decode_symbols function).

Also on Pascal or Volta devices if runned under cuda-memcheck it occasionally detects invalid read access in prefetching code:

```
gpuinflate launch failed: 719
========= Invalid __global__ read of size 4
=========     at 0x000006f8 in GpuInflate.cu:944:prefetch_warp(inflate_state_s volatile *, int)
=========     by thread (93,0,0) in block (0,0,0)
=========     Address 0xb011645cc is out of bounds
=========     Device Frame:GpuInflate.cu:1089:inflate_kernel(gpu_inflate_input_s*, gpu_inflate_status_s*, int) (inflate_kernel(gpu_inflate_input_s*, gpu_inflate_status_s*, int) : 0x3730)

```
**Steps/Code to reproduce bug**

```
nvcc --generate-code arch=compute_52,code=sm_52 -o testGzip kernel.cu GpuInflate.cu -lcuda -lz
run_s.sh
```

or (for second case with cuda-memcheck on Pascal/Volta devices):
```
nvcc --generate-code arch=compute_60,code=[sm_60,sm_70] -o testGzip kernel.cu GpuInflate.cu -lcuda -lz
run_s_chk.sh
```
[gpuinflate.tar.gz](https://github.com/rapidsai/cudf/files/4070679/gpuinflate.tar.gz)

- Environment location: AWS g3.4xlarge (p3.2xlarge) instance

**Environment details**
<details><summary>Click here to see environment details</summary><pre>
     **git***
     Not inside a git repository
     ***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=16.04
     DISTRIB_CODENAME=xenial
     DISTRIB_DESCRIPTION="Ubuntu 16.04.6 LTS"
     NAME="Ubuntu"
     VERSION="16.04.6 LTS (Xenial Xerus)"
     ID=ubuntu
     ID_LIKE=debian
     PRETTY_NAME="Ubuntu 16.04.6 LTS"
     VERSION_ID="16.04"
     HOME_URL="http://www.ubuntu.com/"
     SUPPORT_URL="http://help.ubuntu.com/"
     BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
     VERSION_CODENAME=xenial
     UBUNTU_CODENAME=xenial
     Linux ip-172-31-36-249 4.4.0-1100-aws #111-Ubuntu SMP Wed Dec 4 12:20:15 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
     ***GPU Information***
     Thu Jan 16 10:35:39 2020
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 418.87.01    Driver Version: 418.87.01    CUDA Version: 10.1     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |===============================+======================+======================|
     |   0  Tesla M60           Off  | 00000000:00:1E.0 Off |                    0 |
     | N/A   31C    P0    43W / 150W |      0MiB /  7618MiB |     98%      Default |
     +-------------------------------+----------------------+----------------------+
     +-----------------------------------------------------------------------------+
     | Processes:                                                       GPU Memory |
     |  GPU       PID   Type   Process name                             Usage      |
     |=============================================================================|
     |  No running processes found                                                 |
     +-----------------------------------------------------------------------------+
     ***CPU***
     Architecture:          x86_64
     CPU op-mode(s):        32-bit, 64-bit
     Byte Order:            Little Endian
     CPU(s):                16
     On-line CPU(s) list:   0-15
     Thread(s) per core:    2
     Core(s) per socket:    8
     Socket(s):             1
     NUMA node(s):          1
     Vendor ID:             GenuineIntel
     CPU family:            6
     Model:                 79
     Model name:            Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
     Stepping:              1
     CPU MHz:               2698.996
     CPU max MHz:           3000.0000
     CPU min MHz:           1200.0000
     BogoMIPS:              4600.14
     Hypervisor vendor:     Xen
     Virtualization type:   full
     L1d cache:             32K
     L1i cache:             32K
     L2 cache:              256K
     L3 cache:              46080K
     NUMA node0 CPU(s):     0-15
     Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single kaiser fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt
     ***g++***
     /usr/bin/g++
     g++ (Ubuntu 7.4.0-1ubuntu1~16.04~ppa1) 7.4.0
     Copyright (C) 2017 Free Software Foundation, Inc.
     This is free software; see the source for copying conditions.  There is NO
     warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
     ***nvcc***
     /usr/local/cuda/bin/nvcc
     nvcc: NVIDIA (R) Cuda compiler driver
     Copyright (c) 2005-2018 NVIDIA Corporation
     Built on Sat_Aug_25_21:08:01_CDT_2018
     Cuda compilation tools, release 10.0, V10.0.130
     ***Python***
     /usr/bin/python
     Python 2.7.12
     ***Environment Variables***
     PATH                            : /usr/local/cuda/bin:/home/ubuntu/bin:/home/ubuntu/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
     LD_LIBRARY_PATH                 : /usr/local/cuda/lib64:
     NUMBAPRO_NVVM                   :
     NUMBAPRO_LIBDEVICE              :
     CONDA_PREFIX                    :
     PYTHON_PATH                     :
     conda not found
     ***pip packages***
     /usr/bin/pip
     boto3 (1.9.199)
     botocore (1.12.199)
     docutils (0.14)
     futures (3.3.0)
     jmespath (0.9.4)
     pip (8.1.1)
     Pyste (0.9.10)
     python-dateutil (2.8.0)
     PyYAML (3.11)
     s3transfer (0.2.1)
     setuptools (20.7.0)
     six (1.12.0)
     urllib3 (1.25.3)
     wheel (0.29.0)
You are using pip version 8.1.1, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.</pre>

**Comment by OlivierNV** [comment]
It probably should be added as a comment in gpuinflate.h: but the gpuinflate kernels (as well as other reader kernels) do make the assumption that the address range `[src_addr & ~3 , (src_addr + src_size - 1) | 3]` is accessible, which means that device input allocation size should be rounded up to the next multiple of 4 bytes (otherwise, cuda-memcheck will likely complain about the aligned accesses that cross the unaligned boundary when `src_size % 4 != 0`)

**Comment by isolenta** [comment]
Regarding start address we are fine because it implicitly aligned by cuMemAlloc. For some reason decompression kernel reads few bytes after address range of input buffer (even if size is aligned up to the next multiple of 4). If extend (not only align) size of buffer allocated for the input, cuda-memcheck doesn't complain anymore.

By the way, during decompression of Parquet chunks, their layout in the memory seems doesn't take into account such assumption about last word. So there are may be issues related with chunks layout and decompression.

Hang on Maxwell doesn't depend on address alignment and still here.

**Comment by OlivierNV** [comment]
I'll dig into it, it seems like a bug if we read past `(src+src_size-1)|3` - For the parquet reader (and everything else in cudf), the allocation alignment (both base and size) is guaranteed by rmm (all allocations go through rmm).
Not sure what happens on Maxwell (while cudf is sm_60+, gpuinflate has been tested on Maxwell but with cuda 8.x only) -> do you see the same hang with different cuda compiler versions ?

**Comment by OlivierNV** [comment]
After a quick code inspection, I can indeed see the prefetcher OOB problem when gzip header is specified: the base is adjusted by adding the GZIP header length, but that is not subtracted from src_size, so the prefetcher will read up to `base + src_size + gzip_hdr_len - 8 - 1` instead of the intended `base + src_size - 8 - 1`
Will prepare a PR.

**Comment by isolenta** [comment]
After fix #3821 applied test attached in this issue does not hang on Maxwell anymore.

**Comment by OlivierNV** [comment]
Weird that this was what caused it to hang on Maxwell, but good to know [Closing].
